### PR TITLE
Switch to using a fork of f2c and remove most of _f2c_fixes.py

### PR DESCRIPTION
**Pull request by hoodmane**
I made https://github.com/hoodmane/f2c/ which has some improvements that make it work much better for us including:
1. support for the `recursive` keyword
2. support for variable length arrays that are not arguments in recursive functions
3. `/COMMON/` blocks are labeled with `__attribute__((weak))`
4. automatically omits `ftn_len` arguments from calls to clapack

This allows us to delete most of `_f2c_fixes.py`.

Thanks to @ilayn for writing https://github.com/scipy/scipy/pull/20558, `id_dist` was going to be a huge pain to deal with.

cc @rgommers @agriyakhetarpal

- [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [ ] Add / update tests
- [ ] Add new / update outdated documentation

### CODE REVIEW

**Comment by hoodmane** [comment]
Also, I still am amazed every time I build scipy in 5 minutes. Meson is so great.

File: packages/scipy/patches/0008-MAINT-linalg-Remove-id_dist-Fortran-files.patch
Comment (ryanking13) [inline packages/scipy/patches/0008-MAINT-linalg-Remove-id_dist-Fortran-files.patch:1]: 21,886 lines of diff?? What does this patch do?

File: pyodide-build/pyodide_build/pywasmcross.py [hunk 1]
```diff
@@ -593,7 +594,7 @@ def handle_command(
     if line[0] == "gfortran":
         from pyodide_build._f2c_fixes import replay_f2c
 
-        tmp = replay_f2c(line)
+        tmp = replay_f2c(build_args.f2c_path, line)
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/pywasmcross.py:597]: You removed _f2c_fixes. how do we still call `_f2c_fixes.replay_f2c` here?

File: pyodide-build/pyodide_build/pypabuild.py [hunk 2]
```diff
@@ -270,6 +270,7 @@ def get_build_env(
         args["orig__name__"] = __name__
         args["pythoninclude"] = get_build_flag("PYTHONINCLUDE")
         args["PATH"] = env["PATH"]
+        args["F2C_PATH"] = env.get("F2C_PATH", "f2c")
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/pypabuild.py:273]: I believe passing `F2C_PATH` in the cmd line would work without setting it here, then we won't need to modify `pyodide-build`.

File: packages/scipy/meta.yaml [hunk 3]
```diff
@@ -50,6 +53,9 @@ build:
   # pyodide-build/pyodide_build/_f2c_fixes.py.
   script: |
     set -x
+    git clone https://github.com/hoodmane/f2c.git --depth 1
+    (cd f2c/src && touch gram.c && cp makefile.u makefile && make)
+    export F2C_PATH=$(pwd)/f2c/src/f2c
```
Comment (ryanking13) [inline packages/scipy/meta.yaml:58]: We can remove `f2c` from `requirements.executable`, if we are going to build it here.

**Review by ryanking13 — COMMENTED** [review]
Cool! I've skimmed through your modifications, and TBH I don't understand it very much. Still, I think it is a great improvement if we can remove `_f2c_fixes.py` which is one of our pain point in pyodide-build.

Please ping me again when the PR is ready, I would like to review again.

**Comment by ilayn** [comment]
> Thanks to @ilayn for writing https://github.com/scipy/scipy/pull/20558, id_dist was going to be a huge pain to deal with.

My pleasure. Please let me know if you encounter any bugs because these translations make it extremely easy to make mistakes and I am very good at making those mistakes.

File: packages/scipy/patches/0008-MAINT-linalg-Remove-id_dist-Fortran-files.patch
Comment (hoodmane) [reply packages/scipy/patches/0008-MAINT-linalg-Remove-id_dist-Fortran-files.patch:1]: It's https://github.com/scipy/scipy/pull/20558. It removes the Fortran library called id_dist and replaces it with a reimplementation in cython.

File: pyodide-build/pyodide_build/pywasmcross.py (hunk 1 above)
Comment (hoodmane) [reply pyodide-build/pyodide_build/pywasmcross.py:597]: I didn't remove it, I just deleted 450 out of 662 lines, the remaining 212 lines are still around.

More simplifications can probably be made in a followup but I removed all of the complicated stuff.

File: pyodide-build/pyodide_build/pypabuild.py (hunk 2 above)
Comment (hoodmane) [reply pyodide-build/pyodide_build/pypabuild.py:273]: I tried removing this and it didn't work so I think it's needed. Is there something specific you're suggesting?

Comment (ryanking13) [reply pyodide-build/pyodide_build/pypabuild.py:273]: Hmm, that's strange. Let me check locally.

File: pyodide-build/pyodide_build/pywasmcross.py (hunk 1 above)
Comment (ryanking13) [reply pyodide-build/pyodide_build/pywasmcross.py:597]: Oops. GitHub collapsed the large diff, and I thought you removed it 🙄

File: pyodide-build/pyodide_build/pypabuild.py (hunk 2 above)
Comment (ryanking13) [reply pyodide-build/pyodide_build/pypabuild.py:273]: I opened a PR in your fork: https://github.com/hoodmane/pyodide/pull/9. The build passes while tests fail.

Comment (hoodmane) [reply pyodide-build/pyodide_build/pypabuild.py:273]: Looks like you didn't make the tests fail any worse than they already do though. Thanks!

**Comment by hoodmane** [comment]
@lesteve could you run the scipy tests against this?

**Comment by lesteve** [comment]
It seems like there are Pyodide fatal errors on this PR mostly around `scipy.sparse`, for example when running the equivalent of `pytest --pyargs scipy.sparse.tests.test_array_api`, there is a fatal error (the last output I get is about `test_eigs[bsr]` so maybe this is the culprit?)

```
The cause of the fatal error was:
RuntimeError: null function or function signature mismatch
    at wasm://wasm/000facae:wasm-function[187]:0x8cdf
    at wasm://wasm/0003e6d6:wasm-function[107]:0xbfa0
    at _PyEM_TrampolineCall_JS (/home/lesteve/dev/scipy-tests-pyodide/node_modules/pyodide/pyodide.asm.js:1735:83)
    at pyodide.asm.wasm._PyObject_MakeTpCall (wasm://wasm/pyodide.asm.wasm-029e027e:wasm-function[1136]:0x1c2631)
    at pyodide.asm.wasm.PyObject_Vectorcall (wasm://wasm/pyodide.asm.wasm-029e027e:wasm-function[1145]:0x1c2e37)
    at pyodide.asm.wasm._PyEval_EvalFrameDefault (wasm://wasm/pyodide.asm.wasm-029e027e:wasm-function[3411]:0x29f694)
    at pyodide.asm.wasm._PyEval_Vector (wasm://wasm/pyodide.asm.wasm-029e027e:wasm-function[3412]:0x2a409a)
    at pyodide.asm.wasm._PyFunction_Vectorcall (wasm://wasm/pyodide.asm.wasm-029e027e:wasm-function[1150]:0x1c31e0)
    at pyodide.asm.wasm._PyObject_FastCallDictTstate (wasm://wasm/pyodide.asm.wasm-029e027e:wasm-function[1137]:0x1c2748)
    at pyodide.asm.wasm._PyObject_Call_Prepend (wasm://wasm/pyodide.asm.wasm-029e027e:wasm-function[1154]:0x1c3405) {
  pyodide_fatal_error: true
}
```

The output from running all the scipy tests:
```
--------------------------------------------------------------------------------
Grouped by category
--------------------------------------------------------------------------------
category fatal error or timeout (4 modules)
    scipy.sparse.linalg._eigen.arpack.tests
    scipy.sparse.linalg._eigen.lobpcg.tests
    scipy.sparse.linalg._eigen.tests
    scipy.sparse.tests
category passed (28 modules)
    scipy.cluster.tests
    scipy.constants.tests
    scipy.fftpack.tests
    scipy.fft._pocketfft.tests
    scipy.fft.tests
    scipy.integrate._ivp.tests
    scipy.integrate.tests
    scipy.interpolate.tests
    scipy.io.arff.tests
    scipy.io._harwell_boeing.tests
    scipy.io.matlab.tests
    scipy.io.tests
    scipy._lib.tests
    scipy.linalg.tests
    scipy.misc.tests
    scipy.ndimage.tests
    scipy.odr.tests
    scipy.optimize.tests
    scipy.optimize._trustregion_constr.tests
    scipy.signal.tests
    scipy.sparse.csgraph.tests
    scipy.sparse.linalg._dsolve.tests
    scipy.sparse.linalg._isolve.tests
    scipy.sparse.linalg.tests
    scipy.spatial.tests
    scipy.spatial.transform.tests
    scipy.special.tests
    scipy.stats.tests

--------------------------------------------------------------------------------
Unexpected test results
--------------------------------------------------------------------------------
scipy.sparse.linalg._eigen.arpack.tests result expected in ['passed'], got 'fatal error or timeout' instead
scipy.sparse.linalg._eigen.lobpcg.tests result expected in ['passed'], got 'fatal error or timeout' instead
scipy.sparse.linalg._eigen.tests result expected in ['passed'], got 'fatal error or timeout' instead
scipy.sparse.tests result expected in ['passed'], got 'fatal error or timeout' instead
```

**Comment by hoodmane** [comment]
Thanks @lesteve. I fixed one problem with scipy.sparse, how does it look now?

**Comment by lesteve** [comment]
The remaining issue seems related to `cneupd_`, `call_indirect` is calling it with 25 i32 arguments but its signature has 26 i32 arguments.

**Comment by hoodmane** [comment]
Great, thanks @lesteve!

**Review by ryanking13 — APPROVED** [review]
Codewise looks good to me. Thanks for improving this!

**Comment by agriyakhetarpal** [comment]
I'll rebase #4719 as soon as this is merged, it looks like there are a lot of simplifications, thanks!

**Comment by ilayn** [comment]
In the meantime, we merged MINPACK in https://github.com/scipy/scipy/pull/21131; FYI if that is causing any issues and QUADPACK is next in line, then ARPACK hopefully all in v1.15 timeline.

**Comment by hoodmane** [comment]
We have special workarounds for ARPACK, PROPACK, SUPERLU, fitpack, blas and lapack. Also, there are a bunch of test failures in different parts of scipy but I don't really know were or why.

**Comment by hoodmane** [comment]
id_dist was by far the worst though.

**Pull request merged**